pageserver/ctl/src/main.rs

```diff
@@ -183,6 +190,95 @@ async fn main() -> anyhow::Result<()> {
                 .time_travel_recover(Some(&prefix), timestamp, done_if_after, &cancel)
                 .await?;
         }
+        Commands::Key(DescribeKeyCommand { key }) => {
+            println!("{key:?}");
+
+            macro_rules! kind_query {
+                ($name:ident) => {{
+                    let s: &'static str = stringify!($name);
+                    let s = s.strip_prefix("is_").unwrap_or(s);
+                    let s = s.strip_suffix("_key").unwrap_or(s);
+
+                    #[allow(clippy::needless_borrow)]
+                    (s, pageserver_api::key::$name(key))
+                }};
+            }
+
+            let queries = [
+                ("rel_block", pageserver_api::key::is_rel_block_key(&key)),
+                kind_query!(is_rel_vm_block_key),
+                kind_query!(is_rel_fsm_block_key),
+                kind_query!(is_slru_block_key),
+                kind_query!(is_inherited_key),
+                ("rel_size", pageserver_api::key::is_rel_size_key(&key)),
```

Comment:
why is this not using the macro?

Reply:
Because of the borrow.

Reply:
If the other functions don't have a borrow, then `is_rel_size_key` should not get a borrow either. In general, if the function is inlined there isn't much of a difference, and even if it's not inlined, the ABI often passes large data arguments as reference instead of copying them.

Reply:
I am not sure... I think it might be better to borrow.

> the ABI often passes large data arguments as reference instead of copying them.

I didn't know this, I thought there were bugs which forced references into in-stack copies.

Reply:
all the other functions take it by value, I don't think we should change it for this single function.

Reply:
for the inherent impls point, I do agree that we should do that, it is issue #6438.

Reply:
by-value in [672f9cc](https://github.com/neondatabase/neon/pull/7890/commits/672f9cc6332cd4d36088ff4ea112a53e00d85238)